### PR TITLE
feat(lanelet2_extension): add crosswalk regulatory element

### PR DESCRIPTION
**Pull request by satoshi-ota**
## Description

Add crosswalk regulatory element to lanelet extension library.

The format is [here](https://github.com/orgs/autowarefoundation/discussions/3036).

## Related links

discussion: https://github.com/orgs/autowarefoundation/discussions/3036

## Tests performed

- [x] [PASS TIER IV INTERNAL SCENARIOS](https://evaluation.tier4.jp/evaluation/reports/d2e92155-1f1f-5d62-beb4-5dfa07d0f402?project_id=prd_jt)

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: tmp/lanelet2_extension/include/lanelet2_extension/utility/query.hpp [hunk 1]
```diff
@@ -134,6 +136,13 @@ std::vector<lanelet::NoParkingAreaConstPtr> noParkingAreas(const lanelet::ConstL
  */
 std::vector<lanelet::SpeedBumpConstPtr> speedBumps(const lanelet::ConstLanelets & lanelets);
 
+/**
+ * [speedBumps extracts Crosswalk regulatory elements from lanelets]
```
Comment (mitsudome-r) [inline tmp/lanelet2_extension/include/lanelet2_extension/utility/query.hpp:140]:
```suggestion
 * [crosswalks extracts Crosswalk regulatory elements from lanelets]
```

File: tmp/lanelet2_extension/include/lanelet2_extension/regulatory_elements/crosswalk.hpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  [[nodiscard]] Polygons3d crosswalkArea();
+
+  [[nodiscard]] ConstLineStrings3d stopLines() const;
+  [[nodiscard]] LineStrings3d stopLines();
+
+  [[nodiscard]] ConstLanelet crosswalkLanelet() const;
+
+  /**
+   * @brief add a new no parking area
+   * @param primitive no parking area to add
```
Comment (mitsudome-r) [inline tmp/lanelet2_extension/include/lanelet2_extension/regulatory_elements/crosswalk.hpp:55]:
```suggestion
   * @brief add a new crosswalk area
   * @param primitive crosswalk area to add
```

File: tmp/lanelet2_extension/include/lanelet2_extension/regulatory_elements/crosswalk.hpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+   */
+  void addCrosswalkArea(const Polygon3d & primitive);
+
+  /**
+   * @brief remove a no parking area
+   * @param primitive the primitive
+   * @return true if the no parking area existed and was removed
```
Comment (mitsudome-r) [inline tmp/lanelet2_extension/include/lanelet2_extension/regulatory_elements/crosswalk.hpp:62]:
```suggestion
   * @brief remove a crosswalk area
   * @param primitive the primitive
   * @return true if the crosswalk area existed and was removed
```

File: tmp/lanelet2_extension/include/lanelet2_extension/regulatory_elements/crosswalk.hpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  {
+    return Ptr{new Crosswalk(id, attributes, crosswalk_lanelet, crosswalk_area, stop_line)};
+  }
+
+  /**
+   * @brief get the relevant crosswalk area
+   * @return crosswalk area
+   */
+  [[nodiscard]] ConstPolygons3d crosswalkArea() const;
+  [[nodiscard]] Polygons3d crosswalkArea();
```
Comment (mitsudome-r) [inline tmp/lanelet2_extension/include/lanelet2_extension/regulatory_elements/crosswalk.hpp:46]: It's probably better to make the function name plural since it could return multiple areas.

```suggestion
  [[nodiscard]] ConstPolygons3d crosswalkAreas() const;
  [[nodiscard]] Polygons3d crosswalkAreas();
```

**Review by mitsudome-r — COMMENTED** [review]
Thanks for creating the PR.

Also, please update documentation for the new regulatory element. https://github.com/autowarefoundation/autoware_common/blob/main/tmp/lanelet2_extension/docs/lanelet2_format_extension.md

File: tmp/lanelet2_extension/include/lanelet2_extension/regulatory_elements/crosswalk.hpp (hunk 4 above)
Comment (satoshi-ota) [reply tmp/lanelet2_extension/include/lanelet2_extension/regulatory_elements/crosswalk.hpp]: fixed in https://github.com/autowarefoundation/autoware_common/pull/187/commits/c572b733aaab97f4cc00b8282714deb9be7d4408.

File: tmp/lanelet2_extension/include/lanelet2_extension/regulatory_elements/crosswalk.hpp (hunk 3 above)
Comment (satoshi-ota) [reply tmp/lanelet2_extension/include/lanelet2_extension/regulatory_elements/crosswalk.hpp]: fixed in https://github.com/autowarefoundation/autoware_common/pull/187/commits/c572b733aaab97f4cc00b8282714deb9be7d4408.

File: tmp/lanelet2_extension/include/lanelet2_extension/regulatory_elements/crosswalk.hpp (hunk 2 above)
Comment (satoshi-ota) [reply tmp/lanelet2_extension/include/lanelet2_extension/regulatory_elements/crosswalk.hpp]: fixed in https://github.com/autowarefoundation/autoware_common/pull/187/commits/c572b733aaab97f4cc00b8282714deb9be7d4408.

File: tmp/lanelet2_extension/include/lanelet2_extension/utility/query.hpp (hunk 1 above)
Comment (satoshi-ota) [reply tmp/lanelet2_extension/include/lanelet2_extension/utility/query.hpp]: fixed in https://github.com/autowarefoundation/autoware_common/pull/187/commits/c572b733aaab97f4cc00b8282714deb9be7d4408.

**Comment by satoshi-ota** [comment]
I added documentation in https://github.com/autowarefoundation/autoware_common/pull/187/commits/7ca1cadad14adb4018754f149c4c903b5b4838d9.

**Review by tkimura4 — APPROVED** [review]
LGTM

File: tmp/lanelet2_extension/docs/lanelet2_format_extension.md [hunk 5]
```diff
@@ -202,6 +202,28 @@ The following illustrates how light_bulbs are registered to traffic_light regula
 </relation>
 ```
 
+### Crosswalk
+
```
Comment (mitsudome-r) [inline tmp/lanelet2_extension/docs/lanelet2_format_extension.md:206]: It might be better to add why we are adding this regulatory element in Autoware while original Lanelet2 doesn't require regulatory element.

Example:
```suggestion
Original Lanelet2 format only requires `subtype=crosswalk` tag to be specified in the corresponding lanelet. However, Autoware requires a regulatory element to be defined on top of that in order to:
- explicitly define the relevant driving lanes even in 3D environment
- optionally define stop lines associated with the crosswalk
- enable accurate definition of complex polygons for crosswalk

For the details, refer to this [GitHub discussion](https://github.com/orgs/autowarefoundation/discussions/3036).
```

Comment (satoshi-ota) [reply tmp/lanelet2_extension/docs/lanelet2_format_extension.md:206]: Got it. Thanks for your suggestion!

File: tmp/lanelet2_extension/include/lanelet2_extension/regulatory_elements/autoware_traffic_light.hpp [hunk 6]
```diff
@@ -27,17 +27,17 @@
...
 {
 public:
   using Ptr = std::shared_ptr<AutowareTrafficLight>;
   static constexpr char RuleName[] = "traffic_light";
 
+  struct AutowareRoleNameString
+  {
+    static constexpr const char LightBulbs[] = "light_bulbs";
+  };
+
```
Comment (mitsudome-r) [inline tmp/lanelet2_extension/include/lanelet2_extension/regulatory_elements/autoware_traffic_light.hpp:40]: Is this change intentional?
What is the reason for making this change within this PR?

Comment (satoshi-ota) [reply tmp/lanelet2_extension/include/lanelet2_extension/regulatory_elements/autoware_traffic_light.hpp:40]: I added `AutowareRoleNameString::CrosswalkPolygon`, and moved the struct definition into the class to prevent namespace conflict.

```c++
  struct AutowareRoleNameString
  {
    static constexpr const char CrosswalkPolygon[] = "crosswalk_polygon";
  };
```

Comment (mitsudome-r) [reply tmp/lanelet2_extension/include/lanelet2_extension/regulatory_elements/autoware_traffic_light.hpp:40]: okay, understood.

**Review by mitsudome-r — APPROVED** [review]
LGTM
